pkg/status/sender.go

```diff
@@ -0,0 +1,109 @@
+package status
+
+import (
+	"context"
+	"fmt"
+	"strings"
+	"time"
+
+	"github.com/go-logr/logr"
+	v1 "github.com/llmariner/cluster-manager/api/v1"
+	"google.golang.org/grpc"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/rest"
+)
+
+// New creates a new ComponentStatusSender.
+func New(ctx context.Context, c Config, ns string, opt grpc.DialOption, log logr.Logger) (*ComponentStatusSender, error) {
+	cc, err := grpc.NewClient(c.ClusterManagerServerWorkerServiceAddr, opt)
+	if err != nil {
+		return nil, fmt.Errorf("create cluster manager server worker service client: %s", err)
+	}
+	restConfig, err := rest.InClusterConfig()
+	if err != nil {
+		return nil, err
+	}
+	kc, err := kubernetes.NewForConfig(restConfig)
+	if err != nil {
+		return nil, fmt.Errorf("create kubernetes client: %s", err)
+	}
+	return &ComponentStatusSender{
+		client:       v1.NewClustersWorkerServiceClient(cc),
+		name:         c.Name,
+		namespace:    ns,
+		kc:           kc,
+		logger:       log.WithName("component-status-sender"),
+		interval:     c.Interval,
+	}, nil
+}
+
+// ComponentStatusSender sends component status data to the cluster manager.
+type ComponentStatusSender struct {
+	client       v1.ClustersWorkerServiceClient
+	name         string
+	namespace    string
+	kc           *kubernetes.Clientset
+	logger       logr.Logger
+	interval     time.Duration
+}
+
+// Run starts the component status sender.
+func (s *ComponentStatusSender) Run(ctx context.Context) {
+	s.logger.Info("Starting component status sender...", "name", s.name, "interval", s.interval)
+
+	if err := s.send(ctx); err != nil {
+		s.logger.Error(err, "Failed to send component status message")
+	}
+
+	ticker := time.NewTicker(s.interval)
+	defer ticker.Stop()
+	for {
+		select {
+		case <-ticker.C:
+			if err := s.send(ctx); err != nil {
+				s.logger.Error(err, "Failed to send component status message")
+			}
+		case <-ctx.Done():
+			s.logger.Info("Stopping component status sender...")
+			return
+		}
+	}
+}
+
+func (s *ComponentStatusSender) composeStatusMessage(ctx context.Context) (*v1.ComponentStatus, error) {
```

Comment:
I think it's probably better if a caller defines a status check function. It's a bit odd to have this check in `machine-model-loader` since it cannot check itself (and we need an additional role just for this).

Reply:
sorry, I don't quite understand this comment. Could you explain more?

Reply:
Sorry, what I meant is that this check does not make much sense for `model-manager-loader` since it's a single pod. Checking running pods might make sense for `inference-manager-engine` if the check includes other running model runtime pods.

So I think it's better to let each component define its own way to check the health instead of having the check in the shared package. In particular, if this check is not much useful for `model-manager-loader`, it is not ideal to add a `Role` and `RoleBinding` for that. Generally people want to minimize roles assigned to pods.

Reply:
So, I think it's better if `New` takes a function as one of the arguments and call it from here.

Reply:
Now I see what you mean, 

So basically, for `model-manager-loader` and `session-manager-agent`, no need to send `UpdateComponentStatus`, since we can only return `healthy` and we don't want to include pod status message? For `inference-manager-engine` and `job-manager-dispatcher`, we will send`UpdateComponentStatus` for the pods they created?

Reply:
Yeah that's my current thought. I think we can iterate to improve.

Reply:
Also it's orthogonal, but cluster-manager-server can make the component unhealthy when the health check doesn't come for a long period of time.